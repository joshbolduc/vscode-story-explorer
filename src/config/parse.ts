import { register } from 'esbuild-register/dist/node';
import type { StorybookConfig } from '../storybook/StorybookConfig';
import { getStoriesGlobs } from '../storybook/getStoriesGlobs';
import { hasProperty } from '../util/guards/hasProperty';
import { isNonEmptyString } from '../util/guards/isNonEmptyString';

export const parse = async (configFilePath: string) => {
  register({
    target: `node${process.version.slice(1)}`,
    format: 'cjs',
    hookIgnoreNodeModules: false,
    tsconfigRaw: `{
      "compilerOptions": {
        "strict": false,
        "skipLibCheck": true,
      },
    }`,
  });

  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const result = require(configFilePath) as unknown;
  const config = hasProperty('default')(result) ? result.default : result;

  const stories = hasProperty('stories')(config)
    ? await getStoriesGlobs(config.stories)
    : [];

  const docs = hasProperty('docs')(config) ? config.docs : undefined;

  const autodocs =
    hasProperty('autodocs')(docs) &&
    (typeof docs.autodocs === 'boolean' || docs.autodocs === 'tag')
      ? docs.autodocs
      : undefined;
  const defaultName =
    hasProperty('defaultName')(docs) && isNonEmptyString(docs.defaultName)
      ? docs.defaultName
      : undefined;

  return {
    stories,
    ...(docs
      ? {
          docs: {
            autodocs,
            defaultName,
          },
        }
      : {}),
  } satisfies StorybookConfig;
};
