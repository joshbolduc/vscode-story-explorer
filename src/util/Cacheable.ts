/**
 * An object that lazily performs a predefined operation and caches its result.
 */
export class Cacheable<T> {
  private cachedResult?: { result: T } | undefined;

  /**
   * Initialize a new Cacheable with a given callback.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public constructor(private callback: () => T) {}

  /**
   * Get the result of the operation, invoking it if necessary. If the operation
   * has already been invoked, its result is returned.
   *
   * @returns The result of the operation, cached or otherwise.
   */
  public getResult() {
    if (!this.cachedResult) {
      this.cachedResult = { result: this.callback() };
    }

    return this.cachedResult.result;
  }

  /**
   * Get the cached result of the operation, if any, or a default value if the
   * result has not already been determined.
   */
  public getCachedResult(): T | undefined;
  public getCachedResult<U>(defaultValue: U): T | U;
  public getCachedResult<U>(defaultValue?: U) {
    return this.cachedResult ? this.cachedResult.result : defaultValue;
  }

  /**
   * Replace the callback used to computed the cached result. The cached result,
   * if any, is cleared.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public setCallback(callback: () => T) {
    this.clear();
    this.callback = callback;
  }

  /**
   * Clears the cached result. The next time a result is requested, the
   * operation will be reinvoked and its return value will be cached again.
   */
  public clear() {
    this.cachedResult = undefined;
  }
}
