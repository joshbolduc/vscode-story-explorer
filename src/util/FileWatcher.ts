import type { Disposable, FileSystemWatcher, GlobPattern, Uri } from 'vscode';
import { workspace } from 'vscode';

export type WatchListener = (
  uri: Uri,
  type: 'create' | 'change' | 'delete',
) => void;

export class FileWatcher {
  private fsWatcher?: FileSystemWatcher | undefined;

  private readonly glob: GlobPattern;

  private readonly listenerDisposables: Disposable[] = [];

  public constructor(
    patternOrUri: GlobPattern | Uri,
    callback: WatchListener,
    ignoreCreate = false,
    ignoreChange = false,
    ignoreDelete = false,
  ) {
    this.glob =
      typeof patternOrUri === 'string' || 'pattern' in patternOrUri
        ? patternOrUri
        : patternOrUri.fsPath;
    this.init(callback, ignoreCreate, ignoreChange, ignoreDelete);
  }

  public dispose() {
    this.listenerDisposables.forEach((disposable) => {
      disposable.dispose();
    });
    this.fsWatcher?.dispose();
    this.fsWatcher = undefined;
  }

  private init(
    callback: WatchListener,
    ignoreCreate: boolean,
    ignoreChange: boolean,
    ignoreDelete: boolean,
  ) {
    this.fsWatcher = workspace.createFileSystemWatcher(
      this.glob,
      ignoreCreate,
      ignoreChange,
      ignoreDelete,
    );

    this.listenerDisposables.push(
      this.fsWatcher.onDidChange((uri) => {
        callback(uri, 'change');
      }),
      this.fsWatcher.onDidCreate((uri) => {
        callback(uri, 'create');
      }),
      this.fsWatcher.onDidDelete((uri) => {
        callback(uri, 'delete');
      }),
    );
  }
}
