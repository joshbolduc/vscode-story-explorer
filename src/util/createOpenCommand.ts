import type { TextDocumentShowOptions, Uri } from 'vscode';

interface OpenCommandParams {
  title?: string;
  tooltip?: string;
  arguments: readonly [uri: Uri, options?: TextDocumentShowOptions | undefined];
}

export interface OpenCommandStaticProperties {
  command: 'vscode.open';
}

export const createOpenInEditorCommand = (
  uri: Uri,
  options?: TextDocumentShowOptions,
) => {
  return createOpenCommand({
    title: 'Open',
    tooltip: 'Open in Editor',
    arguments: [uri, options],
  });
};

export const createOpenCommand = <T extends OpenCommandParams>({
  title,
  tooltip,
  arguments: args,
}: T) => {
  return {
    command: 'vscode.open',
    title,
    tooltip,
    arguments: args,
  } as T & OpenCommandStaticProperties;
};
