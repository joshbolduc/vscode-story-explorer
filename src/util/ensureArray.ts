export const ensureArray = <T>(valueOrArray: T | T[]): T[] => {
  if (Array.isArray(valueOrArray)) {
    return valueOrArray;
  }

  return [valueOrArray];
};
