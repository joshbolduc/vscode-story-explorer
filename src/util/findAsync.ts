export const findAsync = async <T>(
  arr: readonly T[],
  predicate: (value: T, index: number, obj: readonly T[]) => Promise<boolean>,
) => {
  for (let i = 0; i < arr.length; i++) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    if (await predicate(arr[i]!, i, arr)) {
      return arr[i];
    }
  }
};
