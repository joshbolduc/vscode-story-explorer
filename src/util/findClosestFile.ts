import { Uri } from 'vscode';
import { findFirstExistingFile } from './findFirstExistingFile';
import { getUriAndParentUris } from './getUriAndParentUris';

export const findClosestFile = (
  filenames: readonly string[],
  startingDir: Uri,
) =>
  findFirstExistingFile(
    getUriAndParentUris(startingDir).flatMap((dir) =>
      filenames.map((filename) => Uri.joinPath(dir, filename)),
    ),
  );
