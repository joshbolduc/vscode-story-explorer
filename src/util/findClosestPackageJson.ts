import type { Uri } from 'vscode';
import { findClosestFile } from './findClosestFile';

export const findClosestPackageJson = (startingDir: Uri) =>
  findClosestFile(['package.json'], startingDir);
