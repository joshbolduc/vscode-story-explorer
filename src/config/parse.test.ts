import { join } from 'path';
import { describe, expect, it } from 'vitest';
import { testBaseDir } from '../../test/util/testBaseDir';
import { parse } from './parse';

describe('parse', () => {
  const files = [
    'project/v6/.storybook/main.js',
    'project/v7/.storybook/main.ts',
    'project/v7/.config-async/main.ts',
    'project/v7/.config-named-exports/main.js',
  ];

  files.forEach((file) => {
    it(`parses ${file}`, async () => {
      expect(await parse(join(testBaseDir, file))).toMatchSnapshot();
    });
  });
});
