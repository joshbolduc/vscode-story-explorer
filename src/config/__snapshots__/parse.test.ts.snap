// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse > parses project/v6/.storybook/main.js 1`] = `
{
  "stories": [
    {
      "directory": "../src/autoTitle",
      "titlePrefix": "Custom title prefix/Second level",
    },
    "../src/**/*.stories.mdx",
    "../src/**/*.stories.@(js|jsx|ts|tsx)",
  ],
}
`;

exports[`parse > parses project/v7/.config-async/main.ts 1`] = `
{
  "docs": {
    "autodocs": true,
    "defaultName": "Autodoc",
  },
  "stories": [
    {
      "directory": "../src/autoTitle",
      "titlePrefix": "Custom title prefix/Second level",
    },
    "../src/features/**/*",
    "../src/stories/**/*.mdx",
    "../src/stories/**/*.stories.@(js|jsx|ts|tsx)",
  ],
}
`;

exports[`parse > parses project/v7/.config-named-exports/main.js 1`] = `
{
  "docs": {
    "autodocs": true,
    "defaultName": "Autodoc",
  },
  "stories": [
    {
      "directory": "../src/autoTitle",
      "titlePrefix": "Custom title prefix/Second level",
    },
    "../src/features/**/*",
    "../src/stories/**/*.mdx",
    "../src/stories/**/*.stories.@(js|jsx|ts|tsx)",
  ],
}
`;

exports[`parse > parses project/v7/.storybook/main.ts 1`] = `
{
  "docs": {
    "autodocs": true,
    "defaultName": "Autodoc",
  },
  "stories": [
    {
      "directory": "../src/autoTitle",
      "titlePrefix": "Custom title prefix/Second level",
    },
    "../src/features/**/*",
    "../src/stories/**/*.mdx",
    "../src/stories/**/*.stories.@(js|jsx|ts|tsx)",
  ],
}
`;
